src/BSON/Document.c

```diff
@@ -471,6 +491,91 @@ static HashTable* php_phongo_document_get_properties(phongo_compat_object_handle
 	return php_phongo_document_get_properties_hash(object, false, 1);
 }
 
+zval* php_phongo_document_read_property(phongo_compat_object_handler_type* object, phongo_compat_property_accessor_name_type* member, int type, void** cache_slot, zval* rv)
+{
+	php_phongo_document_t* intern;
+	char*                  key;
+	size_t                 key_len;
+
+	intern = Z_OBJ_DOCUMENT(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_COMPAT_PROPERTY_ACCESSOR_NAME_TO_STRING(member, key, key_len);
+
+	if (!php_phongo_document_get(intern, key, key_len, rv)) {
+		// Exception already thrown
+		return &EG(uninitialized_zval);
+	}
+
+	return rv;
+}
+
+zval* php_phongo_document_write_property(phongo_compat_object_handler_type* object, phongo_compat_property_accessor_name_type* member, zval* value, void** cache_slot)
+{
+	phongo_throw_exception(PHONGO_ERROR_LOGIC, "Cannot write to %s property", ZSTR_VAL(php_phongo_document_ce->name));
+	return value;
+}
+
+int php_phongo_document_has_property(phongo_compat_object_handler_type* object, phongo_compat_property_accessor_name_type* name, int has_set_exists, void** cache_slot)
+{
+	php_phongo_document_t* intern;
+	char*                  key;
+	size_t                 key_len;
+
+	intern = Z_OBJ_DOCUMENT(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_COMPAT_PROPERTY_ACCESSOR_NAME_TO_STRING(name, key, key_len);
+
+	return php_phongo_document_has(intern, key, key_len);
+}
+
+void php_phongo_document_unset_property(phongo_compat_object_handler_type* object, phongo_compat_property_accessor_name_type* member, void** cache_slot)
+{
+	phongo_throw_exception(PHONGO_ERROR_LOGIC, "Cannot unset %s property", ZSTR_VAL(php_phongo_document_ce->name));
+}
+
+zval* php_phongo_document_read_dimension(phongo_compat_object_handler_type* object, zval* offset, int type, zval* rv)
+{
+	php_phongo_document_t* intern;
+
+	intern = Z_OBJ_DOCUMENT(PHONGO_COMPAT_GET_OBJ(object));
+
+	if (Z_TYPE_P(offset) != IS_STRING) {
+		phongo_throw_exception(PHONGO_ERROR_RUNTIME, "Could not find key of type \"%s\" in BSON data", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(offset));
+		return &EG(uninitialized_zval);
+	}
+
+	if (!php_phongo_document_get(intern, Z_STRVAL_P(offset), Z_STRLEN_P(offset), rv)) {
+		// Exception already thrown
+		return &EG(uninitialized_zval);
+	}
+
+	return rv;
+}
+
+void php_phongo_document_write_dimension(phongo_compat_object_handler_type* object, zval* offset, zval* value)
+{
+	phongo_throw_exception(PHONGO_ERROR_LOGIC, "Cannot write to %s property", ZSTR_VAL(php_phongo_document_ce->name));
+}
+
+int php_phongo_document_has_dimension(phongo_compat_object_handler_type* object, zval* member, int check_empty)
+{
+	php_phongo_document_t* intern;
+
+	intern = Z_OBJ_DOCUMENT(PHONGO_COMPAT_GET_OBJ(object));
+
+	if (Z_TYPE_P(member) != IS_STRING) {
```

Comment:
I didn't see any error tests using unsupported types for the dimension and property accessors. I'm specifically curious about passing an integer, and whether that would be converted to a string or preserved via `zval* member`.

Either way, consider adding some type tests like you have for PackedArray (and split those into their own file).

I don't think this is an issue for the property accessors, since `phongo_compat_property_accessor_name_type` is defined as `zend_string` for PHP 8.0+, but perhaps there's an edge case on PHP 7.4 where it's just a `zval`? I expect that'd probably be a syntax error long before any object handler gets invoked, though -- in which case there's nothing for us to test there.